text/0000-cfg-path-version.md

```diff
@@ -0,0 +1,921 @@
+      This entails that some breakage might sometimes occur when
+      using a nightly compiler.
+
+   2. Answer it in the negative by changing the date when the version constant
+      is bumped in the compiler. That is, a version would only be bumped when
+      releasing new stable or beta compilers and nightly compilers would always
+      be versioned as the latest stable/beta. This also means that given
+      `#[stable(feature = "foobar", since = "1.42.0")]` for some feature
+      `foobar`, the feature would not be available first when the feature
+      actually reaches stable/beta.
+
+   3. As 2. but separate versions reported by `rustc --version` and to
+      `version(..)`. This would for example mean that if the last
+      stable compiler is `1.42.0`, then that would be used by `version(..)`
+      while `rustc --version` would report `1.43.0-nightly`.
+      This approach could be technically achieved by for example
+      maintaining one version constant that tracks the last stable/beta
+      compiler as `x.y.z` and then `--version` would report
+      `x.(y + 1).0-nightly`.
+
+   Two arguments in favour of either 2. or 3. is that they would be more
+   principled as we have not really stabilized something until it reaches
+   stable or beta.
+
+   We consider this unresolved question to be a matter of implementation detail
+   which may be resolved during implementation.
+
+# Possible future work
+[possible future work]: #possible-future-work
+
+## `#[cfg(rust_feature(..))]`
+
+[GAT]: https://github.com/rust-lang/rust/issues/44265
+
+One possible extension we might want to do in the future is to allow users
+to check whether a certain `rustc` feature gate is enabled or not.
+For example, we might write `#[cfg(rustc_feature(generic_associated_types))]`
+to check whether the [GAT] feature is supported in the compiler or not.
+
+The main benefit of such an approach is that it is more direct than checking
+for a particular version. Also note that `clang` uses this approach as noted
+in the [prior art][prior-art].
+
+However, there are some drawbacks as well:
+
+1. The names of feature gates are not always aptly named and usually do not
+   follow a coherent naming system. As a frequent author of RFCs, the author
+   of this one knows that they do not have a principled approach to naming
+   RFCs. The feature name that is then used in the compiler is usually drawn
+   directly from the RFC, so we would either need to accept the random naming
+   of feature gates, or we would need to impose some system.
+
+2. Permitting dependence on the names of feature gates on stable would
+   require us to be more principled with feature gates.
+   For example, `rustc`, or any other Rust compiler, would be unable to
+   remove gates or drastically change their implementations without changing
+   their names. Being more principled could potentially add an undue burden
+   on the library and compiler teams.
```

Comment:
I sense an implicit assumption in these 2 points: that the names used in `rust_feature(...)` _have_ to be the same feature names used in nightly for `#[feature(...)]`.

I would prefer if we treated the latter like everything else in nightly, in a touch-at-your-own-risk sort of way. I imagine that the former ones could be defined more cautiously as we approach stabilization (maybe adhering to some "system" as you mention in point 1).

Maintaining this set of feature identifiers in the long term is not trivial, of course, but I think these 2 drawbacks as stated do not apply if we decouple the two sets of feature names.

Reply:
It seems that if you're defining a stable name for a feature, that feature is now stable and API can't be changed, even if you can only access that feature on nightly. I'm not sure if this was mentioned earlier, but one alternative that might solve the same issue is to have `rust_feature` only work for stable features. You could then use it as an alternative to version detection to know whether you can use a feature on the stable channel, and it doesn't have the downsides of breaking crates that assume the nightly feature will not change before stabilization.

Reply:
There is an implicit assumption here yes. But I feel that assumption is justified.

To have two concurrent set of feature names sounds difficult to maintain and feels highly confusing for users when they need to keep in mind the difference and use `rust_feature(...)`, and I'm already skeptical about the ergonomics of `rust_feature(...)`.

I think there will be a social pressure to pick the name of the unstable feature gate and not change it to something more appropriate out of sheer laziness or to not make things confusing for users.